### DISCUSSION TITLE
failed to load mysqltcl - couldn't load library "libmysqltcl.dll"

### DISCUSSION THREAD

**Discussion by Sbleck** [TPC-Council/HammerDB#241] [Q&A] [answered]
Hello All,

Does anybody have had a problem to use Hammer DB like mentioned below, right after having installed it ?

```
HammerDB CLI v3.2
Copyright (C) 2003-2019 Steve Shaw
Type "help" for a list of commands
The xml is well-formed, applying configuration
hammerdb>librarycheck
Checking database library for Oracle
Success ... loaded library Oratcl for Oracle
Checking database library for MSSQLServer
Success ... loaded library tclodbc for MSSQLServer
Checking database library for Db2
Error: failed to load db2tcl - couldn't load library "C:/HammerDB-3.2/lib/db2tcl2.0.0/libdb2tcl.dll": this library or a dependent library could not be found in library path
Ensure that Db2 client libraries are installed and the location in the PATH environment variable
Checking database library for MySQL
Error: failed to load mysqltcl - couldn't load library "libmysqltcl.dll": this library or a dependent library could not be found in library path
Ensure that MySQL client libraries are installed and the location in the PATH environment variable
Checking database library for PostgreSQL
Error: failed to load Pgtcl - couldn't load library "C:/HammerDB-3.2/lib/pgtcl2.1.1/libpgtcl.dll": this library or a dependent library could not be found in library path
Ensure that PostgreSQL client libraries are installed and the location in the PATH environment variable
Checking database library for Redis
Success ... loaded library redis for Redis
```

I've tried release 4.1, but uninstalled it due to the lack of Redis library.

Please note that the required DLL is there, of course:

```
Microsoft` Windows [versão 10.0.19042.928]
(c) Microsoft Corporation. Todos os direitos reservados.

C:\Users\HP>dir C:\HammerDB-3.2\lib\mysqltcl3.052
 O volume na unidade C é Windows
 O Número de Série do Volume é A62B-0B86

 Pasta de C:\HammerDB-3.2\lib\mysqltcl3.052

08/05/2021  20:41    <DIR>          .
08/05/2021  20:41    <DIR>          ..
19/02/2018  10:25           143.872 libmysqltcl.dll
24/05/2013  09:23               192 pkgIndex.tcl
               2 arquivo(s)        144.064 bytes
               2 pasta(s)   37.471.014.912 bytes disponíveis
```

Any advice should be greatly appreciated.

Thanks in advance,
Sven

**Comment by sm-shaw** [comment, marked as answer]
Hi,
There is a section in the documentation that covers this in detail [Verifying the Installation of Database Client Libraries ](https://www.hammerdb.com/docs/ch01s08.html) Everything you need is on this page including an example walkthrough of finding the MySQL library you are looking for on Windows.

Please note the full error message in librarycheck "this library **or a dependent library** could not be found in library path"
So yes of course libmysqltcl.dll can be found because it is installed with HammerDB but can the dependent libraries be found? In your case HammerDB needs to dynamically load the libmysql.dll library so it needs to be installed and HammerDB needs to be told where to find it.  The documentation explains how to do this. 

Please make sure you read right to the end to section 8.4. MySQL to verify exactly which client library you need for your version of MySQL. Version 3.2 (5.7) and Version 4.1 (8.0) are different.  (Note only MySQL has this dependency on a particular version). 

Finally, for
> I've tried release 4.1, but uninstalled it due to the lack of Redis library. 

Redis has been deprecated due to low single-threaded only performance compared to other databases but can be re-enabled if you wish as unsupported, again [the documentation](https://www.hammerdb.com/docs4.0/ch01s01.html#d0e34) can be a great help.  So as you can see below you can verify that the Redis library is still included with v4.1 and can load it with "package require". 

```
HammerDB CLI v4.1
Copyright (C) 2003-2021 Steve Shaw
Type "help" for a list of commands
The xml is well-formed, applying configuration
hammerdb>librarycheck
Checking database library for Oracle
Success ... loaded library Oratcl for Oracle
Checking database library for MSSQLServer
Success ... loaded library tdbc::odbc for MSSQLServer
Checking database library for Db2
Success ... loaded library db2tcl for Db2
Checking database library for MySQL
Success ... loaded library mysqltcl for MySQL
Checking database library for PostgreSQL
Success ... loaded library Pgtcl for PostgreSQL

hammerdb>package require redis
0.1
hammerdb>
```

Starting with the documentation link above it should be straightforward to install and load the correct libraries.

**Reply by sm-shaw** [reply]
Please mark as answered if this fixed your issue.

**Reply by Sbleck** [reply]
Hello,

Tks for your comments and I'm working on to fix this.

Best Regards,
Sven

**Reply by Sbleck** [reply]
> Please mark as answered if this fixed your issue.

Hello,

Is there any way to diagnose what is happening, when doing benchmarks ?

Sometimes, worked as expected. But, sometimes some error messages appear, even when doing the same steps, mentioned in https://www.hammerdb.com/docs/ch02.html...

Please advice, as soon as possible too.

Tks in adv,
Sven
![Registering some problems during benchmark of Redis database using HammerDB](https://user-images.githubusercontent.com/83974705/118496222-4e7fa700-b6fa-11eb-95ce-5e095191a6dc.jpg)

**Reply by sm-shaw** [reply]
As support/help is not a paid service it is currently on a best endeavours basis without commitment.  In particular for Redis which is deprecated and does not add new features. 
Saying this a quick check shows that it is working even under v4.1 and your error is that you didn't create the Redis schema before trying to test it. 
With Redis please note it is single-threaded without stored procedures so the performance will be a fraction of what other databases can achieve with this type of workload. This is why it is deprecated. 
![red4](https://user-images.githubusercontent.com/38044085/118530922-22394a00-b73d-11eb-9668-f05f8c3d15f2.PNG)
![red1](https://user-images.githubusercontent.com/38044085/118530691-d9819100-b73c-11eb-808d-81a6292de243.PNG)
![red2](https://user-images.githubusercontent.com/38044085/118530708-dc7c8180-b73c-11eb-8ee1-cfd10cc816a5.PNG)
![red3](https://user-images.githubusercontent.com/38044085/118530718-dededb80-b73c-11eb-9042-eefffbd359a6.PNG)

**Comment by Sbleck** [comment]
Hello,

> Saying this a quick check shows that it is working even under v4.1 and
your error is that you didn't create the Redis schema before trying to test
it

Only to say that I waited hours to complete the Redis schema creation, last
saturday:

[image: Concluindo os testes iniciais de benchmark via HammerDB 3.2
realizados no Redis do servidor W2K12 R2 na Amazon EC2 em 15052021.jpg]

> With Redis please note it is single-threaded without stored procedures so
the performance will be a fraction of what other databases can achieve with
this type of workload.

Thanks for your comments, because I've tried to know more about a required
database in some projects, due to be fast (and could work well, in cluster
mode):

[image: Redis vs MongoDB - Which database is right for your business.jpg]

>  This is why it is deprecated

But, for an unemployed guy, noted that a lot of jobs require experience
with this product.

What I wanted was only to know more about concurrency and load capacity,
using it.

Best Regards,
Sven


Em seg., 17 de mai. de 2021 às 14:27, Steve Shaw ***@***.***>
escreveu:

**Reply by sm-shaw** [reply]
Hi,
The load for a small number of warehouses should only take a few minutes and report what it is doing as it is creating them. Note that HammerDB is multithreaded so the more VUs you create given the available CPU cores where HammerDB is installed the more parallel the load will be. The example above took a few minutes.  I don't think you could have run the build correctly. 
Although as noted Redis is single-threaded so achieves performance by running multiple independent instances ie scaling out rather than up.  For database skillsets we would recommend the DB-engines website and ranking here https://db-engines.com/en/ranking.  It is no coincidence that HammerDB focuses on the top 5 relational databases as these are more popular than the other 366 databases combined.
